### PR TITLE
Update util_json

### PR DESCRIPTION
**Pull request by SalahElHazzat**
Added functions to manipulate and normalize jsons in dataset/json.

Entry point main will launch testing code that creates a Json_csv.csv then normalize jsons in dataset/json by creating a new repository dataset/normalized_jsons that contain same hierarchy but json structure is different.

### CODE REVIEW

File: mlmodels/util_json.py [hunk 1]
```diff
@@ -183,4 +183,175 @@ def test_functions_json(arg=None):
...
+import time
+
+
+'''
+    For the given path, get the List of all files in the directory tree
+    Here the Path is where the json directory exist "mlmodels/mlmodels/dataset"
+    Please change path by using os.chdir(***json directory Path***)
+    
+'''
+def getListOfJsonsPaths(dirName):
```
Comment (arita37) [inline mlmodels/util_json.py:198]: Follow my naming convention :

json_xxx_yyyy(
os_folder_getall(

File: mlmodels/util_json.py [hunk 2]
```diff
@@ -183,4 +183,175 @@ def test_functions_json(arg=None):
...
+
 
+'''
+    This function takes as a parameter list of json paths => result of getListOfFiles()
+    Detects unreadable json files and,
+    Returns a list of Python dictionaries of the jsons, containing two keys
+    'Path' key, containing path to the json and 'json' key containing the dictionary
+
+'''
+def Jsons_ToDictionaries(json_Paths):
```
Comment (arita37) [inline mlmodels/util_json.py:224]: No upper case

json_todict()

File: mlmodels/util_json.py [hunk 3]
```diff
@@ -183,4 +183,175 @@ def test_functions_json(arg=None):
...
+                print("Files That have a structure problem:\n")
+            problem+=1
+            print('\t',json_Paths[i])
+            continue
+    print("Total flawed jsons:\t",problem)
+    return data
+
+
+
+'''
```
Comment (arita37) [inline mlmodels/util_json.py:245]: Google
  Docs Strings in python

Or use pycharm editor

File: mlmodels/util_json.py [hunk 4]
```diff
@@ -183,4 +183,175 @@ def test_functions_json(arg=None):
...
+    This function takes as a parameter the result of JsonsToDictionaries,
+    which is a list of dictionaries with 'Path' and 'Json' keys
+    Returns the list of dictionaries 
+    
+'''
+def IndexedDicts_ToDicts(indexed_Dicts):
```
Comment (arita37) [inline mlmodels/util_json.py:251]: Put it as internal function
no need to pollute with too small function

File: mlmodels/util_json.py [hunk 5]
```diff
@@ -183,4 +183,175 @@ def test_functions_json(arg=None):
...
+            Dict[k]=dict()
+        Dict[k]=Update_Dict(l1,Dict[k],value)
+    else:
+        k=Fields_list[0]
+        Dict[k]=value
+        return Dict
+    return Dict
+
+'''transform csv to list of dicts and create new normalized jsons'''
+def csv_toJsons(csv):
```
Comment (arita37) [inline mlmodels/util_json.py:306]: csv_to_json

File: mlmodels/util_json.py [hunk 6]
```diff
@@ -183,4 +183,175 @@ def test_functions_json(arg=None):
...
+    print("New normalized jsons created, check mlmodels\\mlmodels\\dataset")
+    return dicts
+
+###### Testing
+if __name__ == "__main__":
+    JsonsPaths=getListOfJsonsPaths('dataset\\json')
+    Indexed_Dictionaries=Jsons_ToDictionaries(JsonsPaths)
+    Dictionaries=IndexedDicts_ToDicts(Indexed_Dictionaries)
+    Df=IndexedDicts_ToDF(Indexed_Dictionaries)
+    Df.to_csv('Jsons_Csv.csv')
```
Comment (arita37) [inline mlmodels/util_json.py:354]: use

path_norm( “ztmp/myfile”)

ztmp is gitignore
path_norm is mlmodels

**Comment by SalahElHazzat** [comment]
Hi,
I used Pycharm to edit my code and followed your instructions:
- changed functions names as you recommended
- added docstrings to comment my functions
- used path_norm in code execution block

Please let me know if there is any further modifications you want to do.

File: mlmodels/util_json.py [hunk 7]
```diff
@@ -135,52 +138,236 @@ def load_callable_from_uri(uri):
...
+            log(e, p)
 
 
+import json
+import os
+import pandas as pd
+import time
+
+
+def os_folder_getall(folder):
```
Comment (arita37) [inline mlmodels/util_json.py:197]: os_folder_getfiles

File: mlmodels/util_json.py [hunk 8]
```diff
@@ -135,52 +138,236 @@ def load_callable_from_uri(uri):
...
+    for i in range(len(filename)):
+        with open('normalized_jsons\\' + paths_tocreate[i] + '\\' + filename[i] + '.json', 'w') as fp:
+            json.dump(dicts[i], fp, indent=4)
+    print("New normalized jsons created, check mlmodels\\mlmodels\\dataset")
+    return dicts
+
+
+# Testing code
+if __name__ == "__main__":
+    json_path = path_norm("dataset\\json")
```
Comment (arita37) [inline mlmodels/util_json.py:366]: create    test_json_conversion()

File: mlmodels/util_json.py [hunk 9]
```diff
@@ -135,52 +138,236 @@ def load_callable_from_uri(uri):
...
+                if ind < len(lp) - 1:
+                    for j in range(ind + 1, len(lp) - 1):
+                        dire = dire + '\\' + lp[j]
+        paths_tocreate.append(dire)
+    paths_tocreate1 = list(set(paths_tocreate))
+    paths_tocreate1.sort()
+    print("Now creating normalized jsons directory")
+    for p in paths_tocreate1:
+        try:
+            os.mkdir('normalized_jsons' + p)
```
Comment (arita37) [inline mlmodels/util_json.py:353]: os makedirs exist ok=true

File: mlmodels/util_json.py [hunk 10]
```diff
@@ -135,52 +138,236 @@ def load_callable_from_uri(uri):
...
+            d[k] = dict()
+        d[k] = dict_update(l1, d[k], value)
+    else:
+        k = fields_list[0]
+        d[k] = value
+        return d
+    return d
+
+
+def csv_to_skeleton(csv):
```
Comment (arita37) [inline mlmodels/util_json.py:295]: Include this in other function

File: mlmodels/util_json.py [hunk 11]
```diff
@@ -135,52 +138,236 @@ def load_callable_from_uri(uri):
...
+            if problem == 0:
+                print("Files That have a structure problem:\n")
+            problem += 1
+            print('\t', json_paths[i])
+            continue
+    print("Total flawed jsons:\t", problem)
+    return data
+
+
+def dicts_to_df(indexed_dicts):
```
Comment (arita37) [inline mlmodels/util_json.py:244]: Merge all in one function 
root path —-> output : dataframe all merged json.

**Comment by SalahElHazzat** [comment]
Hi,
I made modifications you asked for:
- renamed os_folder_getall(folder) to test_json_conversion(folder)
- created test_json_conversion()
- included usage of os.makedirs() and os.path.exists()
- merged functions as you asked for

**Question:** 
I found that using path_norm( “ztmp/myfile”) is useless in my code, so I removed it.
Can you please launch on your machine to test if code runs correctly?

File: mlmodels/util_json.py [hunk 12]
```diff
@@ -159,28 +159,183 @@ def test_functions_json(arg=None):
...
+            l = fullname[j].split('.')
+            dd = dict_update(l, dd, None)
+        for j in range(len(fullname)):
+            value = ddf.iloc[i][fullname[j]]
+            fields = fullname[j].split('.')
+            dd.update(dict_update(fields, dd, value).copy())
+        dicts.append(dd)
+
+    paths = list(ddf['file_path'])
+    cur_dir=os.getcwd()
```
Comment (arita37) [inline mlmodels/util_json.py:304]: Please use
from mlmodels.util import os_package_root_path
to get current root path

**Comment by SalahElHazzat** [comment]
I updated code to use  os_package_root_path( ) function.

Hope code runs smoothly !

**Comment by SalahElHazzat** [comment]
Hi I added usage of path_norm ( ) in the last commit !

**Pull request merged**